subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterStatisticsTest.groovy

```diff
@@ -0,0 +1,162 @@
+class DirectorySnapshotterStatisticsTest extends Specification {
+    @Rule
+    public final TestNameTestDirectoryProvider tmpDir = new TestNameTestDirectoryProvider(getClass())
+
+    def fileHasher = new TestFileHasher()
+    def statisticsCollector = Mock(DirectorySnapshotterStatistics.Collector)
+    def directorySnapshotter = new DirectorySnapshotter(fileHasher, new StringInterner(), [], statisticsCollector)
+
+    def "can visit missing file"() {
+        when:
+        snapshot(tmpDir.file("missing"))
+
+        then:
+        1 * statisticsCollector.recordVisitHierarchy()
+
+        then:
+        1 * statisticsCollector.recordVisitFileFailed()
+        0 * _
+    }
+
+    def "can visit regular file"() {
+        when:
+        snapshot(tmpDir.createFile("regular.txt"))
+
+        then:
+        1 * statisticsCollector.recordVisitHierarchy()
+
+        then:
+        1 * statisticsCollector.recordVisitFile()
+        0 * _
+    }
+
+    def "can visit empty directory"() {
+        when:
+        snapshot(tmpDir.createDir("dir"))
+
+        then:
+        1 * statisticsCollector.recordVisitHierarchy()
+
+        then:
+        1 * statisticsCollector.recordVisitDirectory()
+        0 * _
+    }
+
+    def "can visit directory hierarchy"() {
+        given:
+        def root = tmpDir.createDir("root")
+        def subDir1 = root.createDir("sub-dir-1")
+        subDir1.createFile("file.txt")
+        def subDir2 = root.createDir("sub-dir-2")
+        subDir2.createFile("file.txt")
+
+        when:
+        snapshot(root)
+
+        then:
+        1 * statisticsCollector.recordVisitHierarchy()
+
+        then:
+        3 * statisticsCollector.recordVisitDirectory()
+        2 * statisticsCollector.recordVisitFile()
+        0 * _
+    }
+
+    @Requires(TestPrecondition.SYMLINKS)
+    def "can visit broken symlink"() {
+        given:
+        def rootDir = tmpDir.createDir("root")
+        rootDir.file('brokenSymlink').createLink("linkTarget")
+        assert rootDir.listFiles()*.exists() == [false]
+
+        when:
+        snapshot(rootDir)
+
+        then:
+        1 * statisticsCollector.recordVisitHierarchy()
+
+        then:
+        1 * statisticsCollector.recordVisitDirectory()
+        1 * statisticsCollector.recordVisitFile()
+        0 * _
+    }
+
+    @Requires(TestPrecondition.SYMLINKS)
+    def "can visit symlinked hierarchy"() {
+        given:
+        def rootDir = tmpDir.createDir("root")
+        def targetDir = rootDir.createDir("target-dir")
+        targetDir.createFile("file.txt")
+        rootDir.file('linked-dir').createLink("target-dir")
+
+        when:
+        snapshot(rootDir)
+
+        then:
+        1 * statisticsCollector.recordVisitHierarchy()
+
+        then:
+        3 * statisticsCollector.recordVisitDirectory()
+        3 * statisticsCollector.recordVisitFile()
+        0 * _
+    }
+
+    @Requires(TestPrecondition.FILE_PERMISSIONS)
+    def "can visit unreadable files"() {
```

Comment:
I don't think we need tests for the unreadable files. Having one test with `recordVisitFileFailed()` should be enough.

Reply:
It actually does seem to count the unreadable directory, so I think it makes sense to keep the test.